3.2 Enable backups over CC transaction protocol

phughk opened the pull request:
Currently we have a separate protocol for backups. This protocol is used in HA but also in the backup tool (neo4j-admin).
Introducing this change enables `neo4j-admin backup` to work with CC endpoints without any change of configuration.

phughk committed:
Reversed dependency between CC and backup

Before this commit, CC is dependant on backup.
Applying this commit will shuffle dependencies around to be more sensible

phughk committed:
Backup client works with transaction protocol

Backup tool works over transaction protocol
CC (core, RR) have the backup protocol disabled PERMANENTLY (switching option won't change anything)
HA backup protocol still supported as it had before

phughk commented on enterprise/backup/src/main/java/org/neo4j/backup/nextgen/Application.java:
```diff
@@ -0,0 +1,96 @@
+/*
```
TODO delete this

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/nextgen/BackupModule.java:
```diff
@@ -0,0 +1,146 @@
+        int maxPages = 1024;
+        PageSwapperFactory pageSwapperFactory = new SingleFilePageSwapperFactory();
+        pageSwapperFactory.setFileSystemAbstraction( fileSystemAbstraction );
+        return new MuninnPageCache( pageSwapperFactory, maxPages, cachePageSize, pageCacheTracer, pageCursorTracerSupplier );
```
Are you intentionally ignoring the page cache configurations in `neo4j.conf`? In doing so, we can no longer do an online backup directly to block device storage, but have to change the procedure to back up to file storage and then import onto block device storage.

phughk replied:
Not intentionally - I will fix that, thanks!

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/TrackingResponseHandler.java:
```diff
@@ -127,6 +127,7 @@ public void onCoreSnapshot( CoreSnapshot coreSnapshot )
 
     Optional<Long> lastResponseTime()
     {
+        System.out.println( "lastResponseTime = " + lastResponseTime );
```
ಠ_ಠ

tinwelint commented:
One thing which makes it a bit harder to read these changes is the lack of javadocs. Class-level docs and on some important methods would do wonders. Cheers.

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupStrategyWrapper.java:
```diff
@@ -0,0 +1,102 @@
+    {
+        wrappedAsRuntimeException( backupStrategyImplementation::init );
+        wrappedAsRuntimeException( backupStrategyImplementation::start );
+        BackupState state = performBackupWithoutLifecycle( onlineBackupContext );
```
You could use `LifeSupport` to take a LifeCycle through its stages instead of doing this wrapping and manual calls, which also doesn't gracefully handles failures along the way.

phughk replied:
Addressed

tinwelint commented on enterprise/backup/src/main/java/org/neo4j/backup/StoreCopyService.java:
```diff
@@ -0,0 +1,79 @@
+        }
+        catch ( Exception e )
+        {
+            throw new RuntimeException( e );
```
`Exceptions.launderedException` instead? This would not wrap `RuntimeException`s one more time.

phughk replied:
Addressed

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/DefaultPageCursorTracerSupplier.java:
```diff
@@ -25,7 +25,7 @@
     public static final DefaultPageCursorTracerSupplier INSTANCE = new DefaultPageCursorTracerSupplier();
 
-    private DefaultPageCursorTracerSupplier()
+    public DefaultPageCursorTracerSupplier()
```
You can just grab the `INSTANCE` where you use this, instead of making the constructor public and creating a new one.

phughk replied:
Addressed

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/OnlineBackupSettings.java:
```diff
@@ -5,19 +5,19 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.backup;
+package org.neo4j.kernel.configuration;
```
Why is it moved to community? Could it not be in, say, enterprise-kernel?

phughk replied:
Addressed

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupCommandArgumentHandler.java:
```diff
@@ -0,0 +1,150 @@
+                    "Perform consistency checks on the label scan store." ) )
+            .withArgument( new OptionalBooleanArg( "cc-property-owners", false,
+                    "Perform additional consistency checks on property ownership. This check is *very* expensive in " +
+                            "time and memory." ) );
```
I think the argument names should be extract to constants, so their definition here won't diverge from the code that looks for them to interpret their meaning. The list of constants would also serve as an easier-to-read overview of what options are available.

chrisvest replied:
I think it would also be desirable to have an option for overriding the page cache memory setting. People run the backup tool on machines that are already running Neo4j instances, more often than they should.

chrisvest replied:
And obviously a page cache size override argument should apply both to the page cache used for the store copy, and the one used for the consistency check.

phughk replied:
Addressed - please have a look at `BackupSupportingClassesFactory` and `HaBackup` for how this has been handled

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupStageOutcome.java:
```diff
@@ -0,0 +1,28 @@
+    SUCCESS,
+    WRONG_PROTOCOL,
+    TRUE_FAILURE,
+    ABSOLUTE_FAILURE;
```
Better names and javadocs would be nice. `SUCCESS` makes sense, and `WRONG_PROTOCOL` you can infer is a failure condition, and you even get an idea about what the problem is. But `TRUE_FAILURE` and `ABSOLUTE_FAILURE` are not very helpful, though.

phughk replied:
kept `TRUE_FAILURE` as is, changed `ABSOLUTE_FAILURE` to `UNRECOVERABLE_FAILURE`, let me know if you have a better suggestion - the idea is that a true failure lets the surrounding class know that what it performed wasn't successful, but it can still try different methods (such as fallback from incremental to full, or fallback from full backup to another strategy). `UNRECOVERABLE_FAILURE` on the other hand means that there was a serious error and there cannot be any fallback.

chrisvest reviewed (changes requested):
I'm only a quarter through it, but I'll give you these comments already since it looks like you are still actively working on it.

martinfurmanski commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupDelegator.java:
```diff
@@ -0,0 +1,95 @@
+ * Simplifies the process of performing a backup over the transaction protocol by wrapping all the necessary classes
+ * and delegating methods to the correct instances.
+ */
+public class BackupDelegator
```
This is not lifecycled and I can't see how catchupClient is used in here.

martinfurmanski commented on enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupContextLoader.java:
```diff
@@ -0,0 +1,62 @@
+        {
+            OnlineBackupRequiredArguments requiredArguments = backupCommandArgumentHandler.establishRequiredArguments( commandlineArgs );
+            Config config = onlineBackupCommandConfigLoader.loadConfig( requiredArguments.getAdditionalConfig() );
+            CheckConsistencyConfig checkConsistencyConfig = backupCommandArgumentHandler.establishConfigurationOverride( config );
```
Perhaps rename?

martinfurmanski commented on enterprise/backup/src/main/resources/META-INF/services/org.neo4j.configuration.LoadableConfig:
```diff
@@ -1 +0,0 @@
-org.neo4j.backup.OnlineBackupSettings
```
delete file?

phughk replied:
Just checked, file was indeed deleted :+1:

martinfurmanski commented on enterprise/backup/src/test/java/org/neo4j/backup/BackupFlowTest.java:
```diff
@@ -0,0 +1,267 @@
+        subject.performBackup( onlineBackupContext );
+    }
+
+    private void anyStrategyPasses()
```
Add annotation or delete test?

phughk replied:
As it turns out, this is used in the above tests to simplify configuring strategy mocks.
I have added a minor java-doc to this method.

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupCommandArgumentHandler.java:
```diff
@@ -0,0 +1,172 @@
+            // We can remove the loading from config file in 4.0
+            if ( arguments.has( "cc-graph" ) )
+            {
+                checkGraph = arguments.getBoolean( "cc-graph" );
```
Many opportunities to use the new constants here.

phughk replied:
Addressed in next push

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupCommandArgumentHandler.java:
```diff
@@ -0,0 +1,172 @@
+    {
+        try
+        {
+            OptionalHostnamePort address = toOptionalHostnamePortFromRawAddress( arguments.parse( args ).get( "from" ) );
```
You can use the constant for `from`

phughk replied:
Addressed in next push

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupCommandArgumentHandler.java:
```diff
@@ -0,0 +1,172 @@
+            long timeout = arguments.get( ARG_NAME_TIMEOUT, TimeUtil.parseTimeMillis );
+            Optional<Path> additionalConfig = arguments.getOptionalPath( ARG_NAME_ADDITIONAL_CONFIG_DIRECTORY );
+            Path reportDir = arguments.getOptionalPath( ARG_NAME_REPORT_DIRECTORY )
+                    .orElseThrow( () -> new IllegalArgumentException( "cc-report-dir must be a path" ) );
```
The exception message can use the relevant constant

phughk replied:
Addressed in next push

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupCopyService.java:
```diff
@@ -0,0 +1,104 @@
+    {
+        try
+        {
+            outsideWorld.fileSystem().renameFile( oldLocation, newLocation );
```
This won't fly on block device. Shame we didn't notice this on the original implementation you copied this from.

phughk replied:
If there is a simple work around I can address it still, that shouldn't be an issue

chrisvest replied:
The `org.neo4j.com.storecopy.StoreCopyClient` has an example of how it could be done. But we are clearly also missing a block device test for the `neo4j-admin backup` command.

chrisvest replied:
The key is figuring out which files are stored on the block device, and then moving those files via the `FileMoveAction.copyViaPageCache`.

phughk replied:
Not addressed - will create a new ticket afterwards.

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupFlowFactory.java:
```diff
@@ -0,0 +1,66 @@
+        List<BackupStrategyWrapper> strategies = Stream
+            .of(
+                new CausalClusteringBackupStrategy( backupDelegator, addressResolutionHelper ),
+                new HaBackupStrategy( haBackup, addressResolutionHelper, onlineBackupContext.getRequiredArguments().getTimeout() ) )
```
The HaBackupStrategy also works when we take online backup of single-instance Enterprise?

phughk replied:
Yes, it is using the backup protocol.

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupStrategyInterface.java:
```diff
@@ -0,0 +1,33 @@
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.lifecycle.Lifecycle;
+
+public interface BackupStrategyInterface extends Lifecycle
```
This doesn't need to outright say `Interface` in the name.

phughk replied:
Addressed in next push

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupStrategyWrapper.java:
```diff
@@ -0,0 +1,127 @@
+
+class BackupStrategyWrapper
+{
+    private final BackupStrategyInterface backupStrategyImplementation;
```
And this field (and associated parameter) doesn't really need to say `Implementation`.

phughk replied:
Addressed in next push

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/HaBackup.java:
```diff
@@ -160,12 +142,11 @@ private BackupOutcome fullBackup( FileSystemAbstraction fileSystem, String sourc
         long timestamp = System.currentTimeMillis();
         long lastCommittedTx = -1;
-        try ( PageCache pageCache = createPageCache( fileSystem, tuningConfiguration ) )
+        try ( PageCache pageCache = this.pageCacheOptional.orElse( createPageCache( fileSystem, tuningConfiguration ) ))
```
I think checkstyle in 3.3. will complain about the lack of a space at the end, there.

chrisvest commented on enterprise/backup/src/main/java/org/neo4j/backup/IdGeneratorWrapper.java:
```diff
@@ -0,0 +1,39 @@
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.kernel.impl.store.id.IdGeneratorImpl;
+
+public class IdGeneratorWrapper
```
This class is awkward because it has no state and no overrides. I would have tested `ClearIdService` with an ephemeral file system instead.

phughk replied:
Do you think it's necessary to change in this PR?

chrisvest commented on enterprise/backup/src/test/java/org/neo4j/backup/OnlineBackupCommandCcIT.java:
```diff
@@ -0,0 +1,228 @@
+    @Test
+    public void backupCanBePerformedOverCcWithDefaultPort() throws Exception
+    {
+        assumeFalse( SystemUtils.IS_OS_WINDOWS );
```
Why do all these tests skip Windows?

phughk replied:
That's what was in the original `OnlineBackupCommandIT` - not sure why

chrisvest commented on enterprise/backup/src/test/java/org/neo4j/backup/OnlineBackupCommandCcIT.java:
```diff
@@ -0,0 +1,228 @@
+
+    @Test
+    @Ignore( "Need to resolved this - causal clustering should have the backup protocol disabled" )
+    public void backupCanNotBePerformedOverHa() throws Exception
```
The test does not look like it's starting an HA cluster.

phughk replied:
The test starts a causal cluster. There are 2 separate IT classes for OnlineBackupCommand:
OnlineBackupCommand**Cc**IT and OnlineBackupCommand**Ha**IT
This test is ignored, because we are expecting the new behaviour for CC to not have the backup protocol enabled (at all). therefore this test should fail.

chrisvest commented on enterprise/backup/src/test/java/org/neo4j/backup/OnlineBackupCommandConfigLoaderTest.java:
```diff
@@ -0,0 +1,153 @@
+public class OnlineBackupCommandConfigLoaderTest
+{
+    @Rule
+    public TemporaryFolder testDirectory = new TemporaryFolder();
```
Why `TemporaryFolder` instead of `TestDirectory`?

chrisvest replied:
I think we really should use `TestDirectory` because that will put the data in `target/test-data`, which on our build agents is more likely to be mounted on a volume that can fit the data, than `TMPDIR` is.

phughk replied:
Addressed in next commit

chrisvest commented on enterprise/backup/src/test/java/org/neo4j/backup/util/TestHelpers.java:
```diff
@@ -0,0 +1,47 @@
+            }
+            return (E) e;
+        }
+        throw new RuntimeException( "The code expected to fail hasn't failed" );
```
Maybe you want to throw `AssertionError`s instead.

phughk replied:
Addressed in next push

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/CoreServerModule.java:
```diff
@@ -136,29 +132,6 @@ public CoreServerModule( IdentityModule identityModule, final PlatformModule pla
-                }
-            } );
-        }
-
```
Is it always enabled, and the extension is registered automatically?

phughk replied:
It should always be disabled for CC (there is no need in maintaining it, since the neo4j-admin will be the same version as the cc running). But you are saying it is registered automatically? That is what I need to avoid (`@Ignore` in OnlineBackupCommandCcIt)

phughk replied:
Resolved - CC (core+RR) no longer have the backup protocol

chrisvest commented on enterprise/kernel/src/test/java/org/neo4j/util/JvmRunner.java:
```diff
@@ -0,0 +1,57 @@
+import org.neo4j.test.ProcessStreamHandler;
+import org.neo4j.test.rule.TestDirectory;
+
+public class JvmRunner
```
Seems pretty backup specific. Maybe just add them as more static methods to `org.neo4j.backup.util.TestHelpers`?

phughk replied:
Because it contains `@Rule TestDirectory` then it is expected to be extended from. The method that doesn't require `NEO4J_HOME` depends on the TestDirectory. Also it had to be moved to enterprise/kernel because there are tests in causal clustering (if I recall) that use this as well. Its a maven tree thing essentially

chrisvest replied:
`OnlineBackupCommandCcIT` is the only test that uses that feature of it, and it could just as well have its own `TestDirectory`.

chrisvest reviewed (changes requested):
Did a second round of review. Still nits to pick.

phughk committed:
Addressed Chris Vest's comments

Moved JvmRunner code to TestHelpers
Moved TestHelpers to enterprise kernel
Resolved CC starting Backup protocol unintentionally.

chrisvest commented:
You got a couple of compilation failures.

phughk commented:
Conflicts are being resolved in linked PR
